File: android/sdk/src/main/kotlin/com/ary/push/internal/token/TokenManager.kt
package com.ary.push.internal.token

import com.ary.push.internal.log.PushLogger
import com.ary.push.internal.storage.StorageManager
import com.ary.push.model.PushProvider
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.coroutines.resume

/**
 * Owns the push token for its whole life: first issue, refresh, replacement, invalidation.
 *
 * A push token is never permanent. FCM rotates it on reinstall, on restore to a new device, on
 * app data clear and occasionally for its own reasons. Every part of the SDK therefore treats
 * the token as a cache of something the transport owns, and the backend is told about every
 * change through the sync queue rather than by the host application.
 */
internal class TokenManager(
    private val storage: StorageManager,
    private val onTokenChanged: (token: String, provider: PushProvider) -> Unit
) {

    private val listeners = CopyOnWriteArrayList<(String) -> Unit>()

    /** Last known token, or null before one has ever been issued. */
    val currentToken: String? get() = storage.getString(StorageManager.KEY_PUSH_TOKEN)

    /** Transport that issued [currentToken]. */
    val provider: PushProvider
        get() = PushProvider.fromWire(storage.getString(StorageManager.KEY_PUSH_PROVIDER))

    fun addListener(listener: (String) -> Unit) {
        listeners.addIfAbsent(listener)
        // A token that arrived before the host application attached its listener is still news
        // to that listener, so it is replayed rather than dropped.
        currentToken?.let { token -> notifyOne(listener, token) }
    }

    fun removeListener(listener: (String) -> Unit) {
        listeners.remove(listener)
    }

    /**
     * Retrieves the token, asking FCM when the SDK has not cached one yet.
     *
     * Returns null rather than throwing when Firebase is unavailable or unconfigured: an
     * application with a missing `google-services.json` should log one clear error and keep
     * running, not crash on launch.
     */
    suspend fun fetchToken(): String? {
        currentToken?.let { return it }
        return requestFromFirebase()
    }

    /** Forces a fresh read from FCM, bypassing the cached value. */
    suspend fun refreshToken(): String? = requestFromFirebase()

    /**
     * Records a token reported by `FirebaseMessagingService.onNewToken`.
     *
     * Idempotent: an unchanged token is ignored, so an OEM that redelivers the same token on
     * every boot does not produce a backend write on every boot.
     */
    @Synchronized
    fun handleNewToken(token: String, provider: PushProvider = PushProvider.FCM) {
        if (token.isBlank()) {
            PushLogger.w { "Ignoring blank push token" }
            return
        }
        val previous = currentToken
        if (previous == token && this.provider == provider) {
            PushLogger.d { "Push token unchanged (${PushLogger.mask(token)})" }
            return
        }

        storage.putString(StorageManager.KEY_PUSH_TOKEN, token, durable = true)
        storage.putString(StorageManager.KEY_PUSH_PROVIDER, provider.wireValue)
        storage.putBoolean(StorageManager.KEY_TOKEN_SYNCED, false)

        PushLogger.i {
            val verb = if (previous == null) "received" else "refreshed"
            "Push token $verb: ${PushLogger.mask(token)} (${provider.wireValue})"
        }

        onTokenChanged(token, provider)
        notifyAll(token)
    }

    /**
     * Deletes the token at the transport.
     *
     * Not part of logout. A user logging out of an application still owns the device and should
     * keep receiving unauthenticated campaigns; deleting the token here would silently
     * unregister the device.
     */
    suspend fun deleteToken(): Boolean = try {
        firebaseMessaging()?.let { messaging ->
            suspendCancellableCoroutine { continuation ->
                messaging.deleteToken().addOnCompleteListener { task ->
                    if (continuation.isActive) continuation.resume(task.isSuccessful)
                }
            }
        } ?: false
    } catch (t: Throwable) {
        PushLogger.e(t) { "Could not delete the push token" }
        false
    }.also { deleted ->
        if (deleted) {
            storage.remove(StorageManager.KEY_PUSH_TOKEN, StorageManager.KEY_PUSH_PROVIDER)
        }
    }

    // ------------------------------------------------------------------ internals

    private suspend fun requestFromFirebase(): String? {
        val messaging = firebaseMessaging() ?: return null
        return try {
            val token: String? = suspendCancellableCoroutine { continuation ->
                messaging.token.addOnCompleteListener { task ->
                    if (!continuation.isActive) return@addOnCompleteListener
                    if (task.isSuccessful) {
                        continuation.resume(task.result)
                    } else {
                        PushLogger.e(task.exception) { "FCM did not return a token" }
                        continuation.resume(null)
                    }
                }
            }
            token?.also { handleNewToken(it, PushProvider.FCM) }
        } catch (t: Throwable) {
            PushLogger.e(t) { "Token request failed" }
            null
        }
    }

    /**
     * Resolves FCM, tolerating an application that has not configured Firebase.
     *
     * `FirebaseMessaging.getInstance()` throws when no `FirebaseApp` has been initialized, which
     * is exactly what happens in an application that added the SDK but has not yet added its own
     * `google-services.json`. That is a configuration mistake with a clear fix, not a crash.
     */
    private fun firebaseMessaging(): FirebaseMessaging? = try {
        FirebaseMessaging.getInstance()
    } catch (t: Throwable) {
        PushLogger.e(t) {
            "Firebase is not initialized in this application, so no push token can be issued. " +
                "Add the host application's own google-services.json and the " +
                "com.google.gms.google-services plugin. See docs/FIREBASE.md."
        }
        null
    }

    private fun notifyAll(token: String) {
        listeners.forEach { listener -> notifyOne(listener, token) }
    }

    private fun notifyOne(listener: (String) -> Unit, token: String) {
        // Host callbacks are foreign code. One that throws must not stop the others, and must
        // never propagate into FCM's service thread.
        runCatching { listener(token) }
            .onFailure { PushLogger.e(it) { "Token refresh listener threw" } }
    }
}

File: ios/Sources/ARYPush/Internal/TokenManager.swift
import Foundation
import UIKit

/// Owns the push token for its whole life: first issue, refresh, replacement, invalidation.
///
/// On iOS the SDK deals with two genuinely different tokens and refuses to conflate them:
///
/// * the **APNs device token**, a hex string Apple hands to the app delegate;
/// * the **FCM registration token**, which Firebase issues only after it has been given the APNs
///   token, and which is what a Firebase-based backend must send to.
///
/// Both are recorded, and ``provider`` says which one the backend should use. An application
/// that uses Firebase Messaging calls ``ARYPush/setFCMToken(_:)`` from
/// `messaging(_:didReceiveRegistrationToken:)`; an application that talks to APNs directly does
/// nothing at all and gets the APNs token.
final class TokenManager {

    private let storage: StorageManager
    private let onTokenChanged: (String, PushProvider) -> Void
    private let lock = NSLock()
    private var listeners: [UUID: (String) -> Void] = [:]

    init(storage: StorageManager, onTokenChanged: @escaping (String, PushProvider) -> Void) {
        self.storage = storage
        self.onTokenChanged = onTokenChanged
    }

    /// The token the backend should send to, given the current provider.
    var currentToken: String? { storage.string(StorageManager.Keys.pushToken) }

    /// The raw APNs device token, kept even when FCM is the active provider.
    var apnsToken: String? { storage.string(StorageManager.Keys.apnsToken) }

    var provider: PushProvider {
        PushProvider.fromWire(storage.string(StorageManager.Keys.pushProvider))
    }

    /// Asks iOS to register with APNs.
    ///
    /// Must run on the main thread, and does nothing useful before the user has authorized
    /// notifications, so callers check authorization first.
    @MainActor
    func registerForRemoteNotifications() {
        UIApplication.shared.registerForRemoteNotifications()
        PushLogger.debug("Requested APNs registration")
    }

    /// Records the APNs device token handed to the app delegate.
    ///
    /// Idempotent: an unchanged token produces no backend write, which matters because iOS
    /// re-delivers the same token on every launch.
    func handleAPNsToken(_ deviceToken: Data) {
        let hex = deviceToken.map { String(format: "%02x", $0) }.joined()
        lock.lock()
        let previous = storage.string(StorageManager.Keys.apnsToken)
        storage.set(StorageManager.Keys.apnsToken, hex)
        lock.unlock()

        if previous != hex {
            PushLogger.info("APNs token \(previous == nil ? "received" : "refreshed"): \(PushLogger.mask(hex))")
        }

        // Firebase, when present, owns the token the backend sends to. Overwriting an FCM token
        // with an APNs token would silently point the backend at the wrong transport.
        guard provider != .fcm else {
            PushLogger.debug("FCM is the active provider; APNs token stored but not synchronised")
            return
        }
        apply(token: hex, provider: .apns)
    }

    /// Records an FCM registration token supplied by the host application.
    func handleFCMToken(_ token: String) {
        guard !token.isEmpty else {
            PushLogger.warn("Ignoring blank FCM token")
            return
        }
        apply(token: token, provider: .fcm)
    }

    /// Reports that APNs registration failed.
    ///
    /// Common and usually not a bug: the Simulator without a paired Mac, a missing entitlement,
    /// or no network at launch. It is logged clearly and never thrown.
    func handleRegistrationFailure(_ error: Error) {
        PushLogger.error(
            "APNs registration failed: \(error.localizedDescription). Check that the Push "
                + "Notifications capability and the aps-environment entitlement are present."
        )
    }

    func addListener(_ listener: @escaping (String) -> Void) -> UUID {
        lock.lock()
        let id = UUID()
        listeners[id] = listener
        let existing = storage.string(StorageManager.Keys.pushToken)
        lock.unlock()

        // A token that arrived before the host attached its listener is still news to it.
        if let existing { deliver(listener, existing) }
        return id
    }

    func removeListener(_ id: UUID) {
        lock.lock()
        listeners[id] = nil
        lock.unlock()
    }

    private func apply(token: String, provider: PushProvider) {
        lock.lock()
        let previousToken = storage.string(StorageManager.Keys.pushToken)
        let previousProvider = storage.string(StorageManager.Keys.pushProvider)
        let unchanged = previousToken == token && previousProvider == provider.wireValue
        if !unchanged {
            storage.set(StorageManager.Keys.pushToken, token)
            storage.set(StorageManager.Keys.pushProvider, provider.wireValue)
        }
        let snapshot = Array(listeners.values)
        lock.unlock()

        guard !unchanged else {
            PushLogger.debug("Push token unchanged (\(PushLogger.mask(token)))")
            return
        }

        PushLogger.info(
            "Push token \(previousToken == nil ? "received" : "refreshed"): "
                + "\(PushLogger.mask(token)) (\(provider.wireValue))"
        )
        onTokenChanged(token, provider)
        snapshot.forEach { deliver($0, token) }
    }

    private func deliver(_ listener: @escaping (String) -> Void, _ token: String) {
        // Host callbacks are foreign code, and they will touch UI.
        DispatchQueue.main.async { listener(token) }
    }
}
